Get `npm install` to work

marcbachmann opened the pull request:
This fixes  #106 and #145

montogeek commented:
@marcbachmann How did you tested it? I just compiled a new version with your changes and the problem persists

montogeek commented:
Also, Could you please fix the dependency version?

rauchg commented on package.json:
```diff
@@ -14,7 +14,8 @@
     "ms": "0.7.1",
-    "uid2": "0.0.3"
+    "uid2": "0.0.3",
+    "user-env": "^2.0.0"
```
Can you pin this please?

rauchg commented:
What an elegant solution! Thank you @marcbachmann. Only one minor nit-pick.

marcbachmann commented:
Done. Rebased and version fixed.

montogeek commented:
Marvelous

marcbachmann commented:
Another solution would be to install `npm` as module and use the api.
That way it's also windows-compatible.

```js
const npm = require('npm')
function install (cb) {
  npm.load({}, function (err) {
    if (err) return cb(err)
    npm.commands.prune([], function (err) {
      if (err) return cb(err)
      npm.commands.install(['--production'], cb)
    })
  })
}
```

rauchg commented:
@marcbachmann I could be wrong but I've heard that the npm team is not recommending API usage. Is this the case @iarna @othiym23 ?

leo commented:
@rauchg [It is.](https://www.youtube.com/watch?v=vLcOkHhEt1g&feature=youtu.be&t=21m1s) Like @othiym23 suggests it, we should a child process instead.

stepankuzmin commented:
![61932189](https://cloud.githubusercontent.com/assets/533564/16900629/c4142230-4c32-11e6-98d0-823f13901a53.jpg)

marcbachmann commented:
I've replaced `user-env` with `shell-env` because it also has windows support and has an async function. I've also removed the prefix that wasn't needed anymore.

@rauchg this is ready to merge.

rauchg commented:
<3

mjgyesme commented:
woo hoo! nvm users rejoice!

iarna commented:
Indeed! We would prefer that you not use npm's internal API, we'd like to keep it subject to change. (Mostly because it's terrible, and we'd like to have the freedom to make it less terrible.)

Calling npm is the correct and supported way of integrating with npm.

If you find yourself having trouble integrating with it (eg it's output is insufficiently uniform, say) then I'd be happy to work with you all in patching it to correct that sort of thing.

rauchg commented:
Thanks a lot for letting us know @iarna
